### ISSUE TITLE
Missing and swapped entries in `DynamicRecordablesMap`

### ISSUE THREAD

**Issue by med-ayssar** [nest/nestml#761]
Assume we have the following simple nestml code using three arrays:

```
neuron simple_model:
  state:
    x [N] real = 1.0
    y [M] real = 2.0
    z [P] real = 3.0
  end
  parameters:
    N integer = 5
    M integer = 3
    P integer = 2
  end
end
```
Having the code generated, we focus only on three components causing this issue:
### State struct
```
  struct State_
  {
enum StateVecVars {
    X = 0,
    Y = 5,
    Z = 8,
};    
    std::vector< double >  x;
    std::vector< double >  y;
    std::vector< double >  z;

    State_();
  };
```
Note: here the enumerator list `x`, `y` and `z` represent the start of each array if we put them next to each other in that order. Notice that `y - z = 3 ` which is the size of the array containing the `y` elements.

###  simple_model::insert_recordables
```
 void simple_model::insert_recordables(size_t first)
  {
      for (size_t i = 0; i < P_.N; i++)
      {
        size_t elem = simple_model::State_::X + i;
        recordablesMap_.insert(get_var_name(i, "X_"), this->get_data_access_functor(elem));
      }
      for (size_t i = 0; i < P_.M; i++)
      {
        size_t elem = simple_model::State_::Y + i;
        recordablesMap_.insert(get_var_name(i, "Y_"), this->get_data_access_functor(elem));
      }
      for (size_t i = 0; i < P_.P; i++)
      {
        size_t elem = simple_model::State_::Z + i;
        recordablesMap_.insert(get_var_name(i, "Z_"), this->get_data_access_functor(elem));
      }
  }
```

### simple_model:get_state_element
```
  inline double get_state_element(size_t elem)
  {
    if(elem >= State_::X && elem < State_::X + P_.N)
    {
      return S_.x[ elem - State_::X];
    }
    else if(elem >= State_::Y && elem < State_::Y + P_.M)
    {
      return S_.y[ elem - State_::Y];
    }
    else
    {
      return S_.z[ elem - State_::Z];
    }
  }
```
Having the vectors in the following order: ` x < y < z`, we can consider having one  `virtual` array having the following form:

| x1 | x2 | x3 | x4 | x5 | y1 | y1 | y2 | z1 | z2 |
|--- |--- |--- |---| ---| --- | --- | ---| --- | ---|    
| 0| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |        

Each instance of the registered model should work fine in nest as long as the arrays keep their size unchanged, but since nest allows modifying the array's size, the following happens:

To demonstrate the issue, we increase the size of the `x` array with one element and decrease the array `y` with one element.
```
n = nest.Create("simple_model", params = {"N": 6, "M": 2})
```
| x1 | x2 | x3 | x4 | x5 | x6 | y1 | y2 | z1 | z2 |
|--- |--- |--- |---| ---| --- | --- | ---| --- | ---|    
| 0| 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 |        

What happens here is that  the arrays have now different sizes than the values stored in the `enum` list.

Now let's consider the `simple_model::insert_recordables` function. The function uses the enumeration list elements as base for each vector to determine the position in the `virtual` array. We have `elem = simple_model::State_::X + i`, for `i in [0, size_of_vector_x - 1]`, and the same is applied for the other vectors. If we show the value of `elem` in each for block, we have the following sets:
```
elem_x = {0, 1, 2, 3, 4, 5 }  // since x.size() == 6 and State_::X = 0

elem_y = {5, 6} // since x.size() == 2 and State_::Y = 5. Warning: as you can see in the second table above, y starts at 6

elem_z = {8, 9}
```

The Entry 5 is covered twice, and it is registered twice in the map under two different names (`y1` and  `x6`) (recordable names have `i+ 1` as suffix). On the other hand, the Entry 7 isn't covered at all, it should be covered by the `elem_y` set, because the ` State_::Y = 5` doesn't sync with the array size.

The second problem is in the `get_state_element` function. What the function does is just mapping a global position to a local position in the `virtual` array (see table) and extract the value for one of the vectors.

Since the `virtual` array spans over 10 entries, we can easily map by hand each global position into its local position and see which value it holds.

let 's first define the mapping logic used in the function:
```
CASE1:
elem >= State_::X && elem < State_::X + P_.N => x[ elem - State_::X]

CASE2:
elem >= State_::Y && elem < State_::Y + P_.M => y[ elem - State_::Y]

ELSE:
 z[ elem - State_::Z]
```

| elem        | local          | actual value  | expected | 
| ------------- |:-------------:| ------------:| ---------------:|
| 0      | 0 | x1 | x1| 
| 1     | 1 | x2| x2| 
| 2    | 2| x3| x3| 
| 3      | 3 | x4 | x4| 
|4     | 4 | x5| x5| 
| 5     | 5 | x6 | x6| 
| 6    | 1 | **y2**| y1| 
| 7     | 1 | **undefined** | y2| 
| 8  | 0 | z1| z1| 
| 9     | 1 | z2 | z2| 

Only the global entries 6 and 7 are of interests. To compute the local position of 6 we apply the following:
-  6 doesn't fulfill `CASE1`, because (6 >=0 and 6 < 0 + 6) is `FALSE´
- 6 fulfills `CASE2`, because (6 >= 5 and 6 < 5 + 2) is `True` ⇾  6 is mapped to `6 - 5 = 1`, which is the second element in the `y` vector, but it must be the **first** element.

For 7:
-  7 doesn't fulfill `CASE1`, because (7>=0 and 7 < 0 + 6) is `FALSE´
- 7 doesn't fulfill `CASE2`, because (7>=5 and 7 < 5 + 2) is `FALSE´
- 7 is then mapped to `7 - 8 = -1`. Reading an element from the `z` vector returns a **garbage value**, mainly an undefined behavior. 

To summarize, registering vectors in the `DynamicRecordablesMap` isn't fully correct, as we have missing array entries are ignored, some are swapped and others might hold undefined values.
